File: 백준/Silver/28215. 대피소/대피소.py
import sys
from itertools import combinations

n, k = map(int, sys.stdin.readline().split())
homes = [tuple(map(int, sys.stdin.readline().split())) for _ in range(n)]
maximum_dist = []

for shelter in combinations(homes, k):
    if n != k:
        n_shelter = list(set(homes) - set(map(tuple, shelter))) # 대피소가 아닌 집들
        
    else:
        n_shelter = shelter
        
    distances = []
    
    for i in n_shelter:
        min_dist = 10000000000
        for j in shelter:
            dist = (abs(i[0]-j[0]) + abs(i[1]-j[1]))
            min_dist = min(min_dist, dist)
        distances.append(min_dist)
    
    maximum_dist.append(max(distances))

print(min(maximum_dist))

File: 백준/Silver/28215. 대피소/대피소.cc
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;
using vi = vector<int>;

vector<int> x;
vector<int> y;
int n;
int k;

int calc(int p1, int p2, int p3) {
    int ans = 0;
    for (int i = 0; i < n; ++i) {
        ans = max(ans, min({ abs(x[i] - x[p1]) + abs(y[i] - y[p1]), abs(x[i] - x[p2]) + abs(y[i] - y[p2]), abs(x[i] - x[p3]) + abs(y[i] - y[p3]) }));
    }
    return ans;
}

int main()
{
    cin >> n >> k;
    vector<int> xx(n);
    vector<int> yy(n);
    x = xx;
    y = yy;

    int ans = 99999999;
    for (int i = 0; i < n; ++i)
    {
        int a, b;
        cin >> a >> b;
        x[i] = a;
        y[i] = b;
    }

    if (k == 1) {
        for (int i = 0; i < n; ++i)
        {
            ans = min(ans, calc(i, i, i));
        }
    }
    else if (k == 2) {
        for (int i = 0; i < n; ++i) {
            for (int j = i; j < n; ++j) {
                ans = min(ans, calc(i, j, j));
            }
        }
    }
    else if (k == 3) {
        for (int i = 0; i < n; ++i) {
            for (int j = i; j < n; ++j) {
                for (int k = j; k < n; ++k) {
                    ans = min(ans, calc(i, j, k));
                }
            }
        }
    }
    cout << ans << '\n';

    return 0;
}